### PR TITLE
feat(storage): add StartOffset and EndOffset to Query

### LINKED ISSUE #2556: storage: support startOffset query during object listing
**Is your feature request related to a problem? Please describe.**
If we have a large number of files stored in the bucket and we only want to list objects starting from a certain prefix, listing objects starting from a specific starting point (under a lexicographical order) would be desirable. This[ feature is supported in the JSON API via `startOffset`](https://cloud.google.com/storage/docs/json_api/v1/objects/list). However, I could not find a way to set this startOffset value under the go sdk.

**Describe the solution you'd like**
Support for `startOffset` when listing objects should be simple. This can be done by:
1. Add `StartOffset` field to `storage.Query` struct
2. Request changes to internal storage client api to add `func (c *BucketsListCall) StartOffset(startOffset string)` to modify the url query parameter for startOffset
3. Modify `func (it *ObjectIterator) fetch(pageSize int, pageToken string)` to incorporate startOffset parameter if it is being set in the Query, ie. `req.StartOffset(it.query.StartOffset)`

To use this feature, I should be able to set the `StartOffset` field in `storage.Query` to some string value, and expect the internal listObjects call to respect the startOffset and list objects starting from startOffset.

### PR DESCRIPTION
**Pull request by AlisskaPie**
Closes #2556

### CODE REVIEW

**Comment by AlisskaPie** [comment]
@tritone Do we need to add a test here? If so, where it will be better to add this test?
I've tested it in golang-samples repo, and it works correctly.

**Comment by frankyn** [comment]
Hi @tritone, yes these fields are okay to surface as they're published in public documentation: https://cloud.google.com/storage/docs/json_api/v1/objects/list

File: storage/storage.go [hunk 1]
```diff
@@ -1297,6 +1297,18 @@ type Query struct {
 	// the query. It's used internally and is populated for the user by
 	// calling Query.SetAttrSelection
 	fieldSelection string
+
+	// StartOffset sets the optional parameter "startOffset": Filter results
```
Comment (tritone) [inline storage/storage.go:1301]: No need for the "sets the optional parameter" verbiage. I would say:

`StartOffset is used to filter results to objects whose names are...` etc.

File: storage/storage.go [hunk 2]
```diff
@@ -1297,6 +1297,18 @@ type Query struct {
...
+	// to objects whose names are lexicographically equal to or after
+	// startOffset. If endOffset is also set, the objects listed will have
+	// names between startOffset (inclusive) and endOffset (exclusive).
+	StartOffset string
+
+	// EndOffset sets the optional parameter "endOffset": Filter results to
```
Comment (tritone) [inline storage/storage.go:1307]: Same as above.

File: storage/storage.go (hunk 1 above)
Comment (AlisskaPie) [reply storage/storage.go:1301]: > I think an integration test might make sense for this feature-- something like
> 
> https://github.com/googleapis/google-cloud-go/blob/2a870c27ead0e5f70b8290079fce1e5005e4f427/storage/integration_test.go#L1055
> 
> and the tests below that. @AlisskaPie could you create this?

Yes, I'll do this.
Thank you for the detailed answer!

**Comment by AlisskaPie** [comment]
> I think an integration test might make sense for this feature-- something like
> 
> google-cloud-go/storage/integration_test.go
> 
> Line 1055 in 2a870c2
> 
>  func testObjectIterator(t *testing.T, bkt *BucketHandle, objects []string) { 
> and the tests below that. @AlisskaPie could you create this?
> 

Yes, I'll do this.
Thank you for the detailed answer!

File: storage/storage.go (hunk 1 above)
Comment (AlisskaPie) [reply storage/storage.go]: Done

File: storage/integration_test.go [hunk 3]
```diff
@@ -1069,13 +1069,42 @@ func testObjectIterator(t *testing.T, bkt *BucketHandle, objects []string) {
 	if !ok {
 		t.Errorf("ObjectIterator.Next: %s", msg)
 	}
+	m := make(map[string][]*ObjectAttrs)
```
Comment (tritone) [inline storage/integration_test.go:1072]: These look like the right things to test. Could you move this into its own testing function, call it `testObjectIteratorWithOffset` or similar? That way we can easily see that a failure is due to offset issues.

File: storage/integration_test.go [hunk 4]
```diff
@@ -1069,13 +1069,42 @@ func testObjectIterator(t *testing.T, bkt *BucketHandle, objects []string) {
...
+		}
+	}
 	// TODO(jba): test query.Delimiter != ""
 }
 
 func testObjectsIterateSelectedAttrs(t *testing.T, bkt *BucketHandle, objects []string) {
 	// Create a query that will only select the "Name" attr of objects, and
-	// invoke object listing.
-	query := &Query{Prefix: ""}
+	// invoke object listing, with given offsets. The last object will be excluded.
```
Comment (tritone) [inline storage/integration_test.go:1102]: Let's not have this test include StartOffset/EndOffset just to have a variety of queries (you can leave it in the test below though).

File: storage/integration_test.go (hunk 3 above)
Comment (AlisskaPie) [reply storage/integration_test.go:1087]: Done

File: storage/integration_test.go (hunk 4 above)
Comment (AlisskaPie) [reply storage/integration_test.go]: Okay, I left it in the testObjectsIterateAllSelectedAttrs

**Pull request merged**